history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
+	if nextEp < h.NextEpoch() {
+		return ErrBadEpoch
+	}
+	nextStr := h.currentSTR.generateNextSTR(m, nextEp)
+	h.currentSTR = nextStr
+	return nil
+}
+
+func (h *History) LookUp(key string) (MerkleNode, *AuthenticationPath) {
+	str := h.currentSTR
+	return str.tree.Get(key)
+}
+
+func (h *History) LookUpInEpoch(key string, ep int64) (MerkleNode, *AuthenticationPath) {
+	str := h.GetSTR(ep)
+	return str.tree.Get(key)
+}
+
+func (h *History) GetSTR(ep int64) *SignedTreeRoot {
+	pointer := h.currentSTR
+	for pointer.epoch > ep && pointer != nil {
```

Comment:
> Also, what happens if GetSTR is given an ep that's greater than the current epoch? Does it make sense to return the current str, as it's doing? Or, should we err? Similarly, if it's less than the starting epoch (since that's a parameter to the NewHistory).

Yes, I also think about it. May @masomel know how to handle these situations? The current flow is adopted from the reference implementation.